### DISCUSSION TITLE
TypeScript Doesn't Read Type Definition Files?

### DISCUSSION THREAD

**Discussion by cemalgnlts** [CodinGame/monaco-vscode-api#316] [Q&A] [answered]
Hi,

I created the following structure on the demo page:
```
tmp/
├── node_modules/
│   ├── @types/
│   │   └── logger/
│   │       └── index.d.ts
│   └── logger/
│       └── index.js
└── main.ts
```

My expectation was that when I imported the module named `logger`, it would find the type definition from the `@types` folder but it didn't work as I hoped.

**node_modules/@types/logger/index.d.ts**
```ts
declare class Logger {
    constructor(name: string);
    info(msg: string): void;
}
```

**node_modules/logger/index.js**
```js
class Logger {
    constructor(name) {}
    info(msg) {}
}

export default Logger;
```

**main.ts**
```ts
import Logger from "logger";

const logger = new Logger();
logger.// No autocomplete.
```

Did I miss something?

**Comment by CGNonofr** [comment]
Does it work in VSCode?

**Reply by cemalgnlts** [reply]
No :)

I realized that the type file should be like this:

**node_modules/@types/logger/index.d.ts**
```ts
declare module "logger" {
  export default class Logger {
    constructor(name: string);
    info(msg: string): void;
  }
}
```

Now it works in VSCode but not in the sample page. You can see that VSCode successfully reads the type file and warns me for not giving the name argument, it also warns me for not giving an argument to a function. But on the sample page TS does not care about any of this.

| VSCode | Demo Page |
|--------|-----------|
| ![Screenshot 2024-01-16 16 22 19](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/a0438691-d03b-4fa7-9357-bba063426277) | ![Screenshot 2024-01-16 16 24 08](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/f13f9ebd-283e-4c09-be71-452a613b976b) |

**Reply by CGNonofr** [reply]
Can you try on vscode.dev? that's the closer to what is the expected behavior here

**Reply by cemalgnlts** [reply]
It doesn't work on vscode.dev & github.dev, it works on GitHub Code Spaces and Gitpod.

**Reply by cemalgnlts** [reply]
I didn't notice it before, but there is this warning on the bottom right of the sample page:

![partial-mode-status-item](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/c65bb4b8-3a04-47cf-a684-63e1e5731c9f)

https://code.visualstudio.com/docs/nodejs/working-with-javascript#_partial-intellisense-mode

**Reply by CGNonofr** [reply]
Did you change the server.host on the vite config? the typescript worker is only able to enable project wide intellisense if the assets domain is different from the main domain

**Reply by cemalgnlts** [reply]
Project Wide IntelliSense problem is only on the demo page (https://codingame.github.io/monaco-vscode-api/), not on the project I created, but it still doesn't care about type definitions.

I found some errors in the console:

| First Errors | Next Errors |
|--------------|-------------|
| ![Screenshot 2024-01-16 19 30 01](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/91a79d6f-27e1-4bd5-91df-f2fc37171f5e) | ![Screenshot 2024-01-16 19 31 34](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/f86ddd0e-91ed-483c-8aeb-e73be61b43b5) |

**Reply by CGNonofr** [reply]
I think the typescript project wide intellisense done client-side is still in "alpha" on the VSCode side. vscode.dev doesn't use it because it's only enabled when the ext-host iframe is served on another domain, which is not the case currently on vscode.dev

**Reply by cemalgnlts** [reply, marked as answer]
I found a strange solution. The folder structure should be like this:

```
node_modules/
└── logger/
    ├── dist/
    │   └── logger.d.ts
    ├── index.js
    └── package.json
```

package.json should be like this:
```json
{
  "name": "logger",
  "version": "1.0.0",
  "types": "./dist/logger.d.ts"
}
```

There must be a dist folder, the name of the type file must be the same as the package file.

I also found out why you don't detect semantic errors. I changed the `typescript.tsserver.web.projectWideIntellisense.suppressSemanticErrors` setting to `false` and now autocomplete works and semantic errors are also shown.

![Screenshot 2024-01-18 16 30 46](https://github.com/CodinGame/monaco-vscode-api/assets/45357531/7026b22a-045d-4ada-8929-3e36378a0874)

Thanks!
